### ISSUE TITLE
Still may have duplicated RUV after reindexing or importing in a task although nsuniqueid index is OK.

### ISSUE THREAD

**Issue by progier389** [389ds/389-ds-base#4934]
**Issue Description**
Although the symptoms are the same, this bug is different from  https://github.com/389ds/389-ds-base/issues/4910:

The Replication may try to update RUV while a reindex or an import task is running:

1.   The mapping tree is rightly set in read-only mode during the task
2.  but ruv update is flagged as OP_FLAG_TOMBSTONE_ENTRY and
3.  mtn_get_be ignore the referrals when this flag is set.  (to be able to delete tombstone on consumers)
4.  So the modify operation continue until the dn is not found because entryrdn is under construction and return LDAP_NO_SUCH_OBJECT object error.
5.  But if we get this error the ruv update function considers that there is no RUV and creates a new one
6. and the add operation succeed as there is no DN control on tombstone (furthermore the write txn serialization may 
     postpone the real add operation until the reindex get completed)
    
 **Proposed solution:**
 Check that we have both OP_FLAG_TOMBSTONE_ENTRY flag and SLAPI_OPERATION_DELETE operation type
  rather than simply checking  OP_FLAG_TOMBSTONE_ENTRY flag
  to override the return of referrals
 so the error code will either be LDAP_REFERRAL or LDAP_UNWILLING_TO_PERFORM and the RUV update will be aborted 
  whitout trying to create a new RUV

**Package Version and Platform:**
 - Platform: Fedora
 - Package and version: [e.g. 389-ds-base-1.4.4.4-20200721git5d41dc5a4.fc32.x86_64]

**Steps to Reproduce**
I reproduced systematically the issue while running test_ruv_after_reindex pytest on lmdb:
export NSSLAPD_DB_LIB=mdb
cd dirsrvtests/tests/suites/replication
pytest $PWD -k test_ruv_after_reindex

**Expected results**
Test should PASS

Note: I suspect that the same ssue could occurs with bdb but it is less frequent and should be fixed in same release than https://github.com/389ds/389-ds-base/issues/4910

**Comment by progier389** [comment]
I discovered the reason the test was failing with lmdb (it is not related to this issue but  to a bug in dbmdb_db2index
Anyway the tests done to understand this bg showed that the issue described above is still possible but within a very limited time window:
  The problem may only occurs while reindexing (in import case the backend is in full referral mode) 
  The problem occurs if the search occurs while the reindex is running but the add occurs once the reindex has finished 
  The proposed solution could not work as we need to be able to write the RUV on consumer
  I was able to think about two solutions: 
1. Preventing backend status change while replica_write_ruv is running: Add a new flag (REINDEXING) in backend flags. Lock backend in read only mode while running replica_write_ruv and check the REINDEXING flags- the db2index function should set the flag (protected by backend write lock)
2. Set the backend in full referral mode if one of entrydn / entryrdn / nsuniqueid is reindexed (as the search may returns inconsistent data we should not left the backend in read only mode anyway)

**Comment by Firstyear** [comment]
I think the backend status change flag is more flexible and extends an already existing state machine, so I think that's a better way to proceed.

**Comment by progier389** [comment]
I agree. That solution has also the advantage of been easily understandable that we are protected against race condition.
On the other side I think that we also should implement the full referral mode because we could not let search return inconsistent data.  So I think that we should implement both solutions ...

**Comment by Firstyear** [comment]
Both is better than none, that's for sure!

**Comment by mreynolds389** [comment]
No longer reproducible, closing
